// material-community-components/esm2015/scrollspy/scrollspy.directives.js
import * as tslib_1 from "tslib";
import { Input, ContentChildren, Directive, ElementRef, QueryList, Renderer2, } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { MccScrollspyService } from './scrollspy.service';
let MccScrollspyItemDirective = class MccScrollspyItemDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    /**
     * Hold the element id, if element doesn't have id
     * the method will create one
     */
    set id(id) {
        if (!id) {
            id = this._createId();
            this.renderer.setProperty(this.elementRef.nativeElement, 'id', id);
        }
        this._id = id;
    }
    get id() {
        return this._id;
    }
    /**
     * Element distance of the top
     */
    get top() {
        return this.elementRef.nativeElement.offsetTop;
    }
    /**
     * Element is focused
     */
    set focus(focused) {
        this._focused = coerceBooleanProperty(focused);
    }
    get focus() {
        return this._focused;
    }
    ngAfterContentInit() {
        if (!this.label) {
            this.label = this.elementRef.nativeElement.textContent;
        }
        this.id = this.elementRef.nativeElement.id;
    }
    /**
     * Create an ID for the element
     */
    _createId() {
        let tmpID = this.label.toLowerCase().replace(/[ ]+/gi, '_');
        return `mcc_scrollspy_${tmpID}`;
    }
};
tslib_1.__decorate([
    Input('focus'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], MccScrollspyItemDirective.prototype, "focus", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], MccScrollspyItemDirective.prototype, "label", void 0);
MccScrollspyItemDirective = tslib_1.__decorate([
    Directive({
        selector: '[mccScrollspyItem], [mcc-scrollspy-item]',
        exportAs: 'mccScrollspyItem',
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], MccScrollspyItemDirective);
export { MccScrollspyItemDirective };
let MccScrollspyGroupDirective = class MccScrollspyGroupDirective {
    constructor(mccScrollspyService) {
        this.mccScrollspyService = mccScrollspyService;
    }
    /**
     * Name of the scrollspy group
     */
    set name(name) {
        this._name = name;
    }
    ngAfterContentInit() {
        // add the group items
        const items = this.items.map(item => item);
        this.mccScrollspyService.create(this._name, items);
    }
};
tslib_1.__decorate([
    ContentChildren(MccScrollspyItemDirective),
    tslib_1.__metadata("design:type", QueryList)
], MccScrollspyGroupDirective.prototype, "items", void 0);
tslib_1.__decorate([
    Input('mccScrollspyGroup'),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], MccScrollspyGroupDirective.prototype, "name", null);
MccScrollspyGroupDirective = tslib_1.__decorate([
    Directive({
        selector: '[mccScrollspyGroup], [mcc-scrollspy-group]',
        exportAs: 'mccScrollspyGroup',
    }),
    tslib_1.__metadata("design:paramtypes", [MccScrollspyService])
], MccScrollspyGroupDirective);
export { MccScrollspyGroupDirective };
