// material-community-components/esm5/scrollspy/scrollspy.directives.js
import * as tslib_1 from "tslib";
import { Input, ContentChildren, Directive, ElementRef, QueryList, Renderer2, } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { MccScrollspyService } from './scrollspy.service';
var MccScrollspyItemDirective = /** @class */ (function () {
    function MccScrollspyItemDirective(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    Object.defineProperty(MccScrollspyItemDirective.prototype, "id", {
        get: function () {
            return this._id;
        },
        /**
         * Hold the element id, if element doesn't have id
         * the method will create one
         */
        set: function (id) {
            if (!id) {
                id = this._createId();
                this.renderer.setProperty(this.elementRef.nativeElement, 'id', id);
            }
            this._id = id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MccScrollspyItemDirective.prototype, "top", {
        /**
         * Element distance of the top
         */
        get: function () {
            return this.elementRef.nativeElement.offsetTop;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MccScrollspyItemDirective.prototype, "focus", {
        get: function () {
            return this._focused;
        },
        /**
         * Element is focused
         */
        set: function (focused) {
            this._focused = coerceBooleanProperty(focused);
        },
        enumerable: true,
        configurable: true
    });
    MccScrollspyItemDirective.prototype.ngAfterContentInit = function () {
        if (!this.label) {
            this.label = this.elementRef.nativeElement.textContent;
        }
        this.id = this.elementRef.nativeElement.id;
    };
    /**
     * Create an ID for the element
     */
    MccScrollspyItemDirective.prototype._createId = function () {
        var tmpID = this.label.toLowerCase().replace(/[ ]+/gi, '_');
        return "mcc_scrollspy_" + tmpID;
    };
    tslib_1.__decorate([
        Input('focus'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], MccScrollspyItemDirective.prototype, "focus", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], MccScrollspyItemDirective.prototype, "label", void 0);
    MccScrollspyItemDirective = tslib_1.__decorate([
        Directive({
            selector: '[mccScrollspyItem], [mcc-scrollspy-item]',
            exportAs: 'mccScrollspyItem',
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
    ], MccScrollspyItemDirective);
    return MccScrollspyItemDirective;
}());
export { MccScrollspyItemDirective };
var MccScrollspyGroupDirective = /** @class */ (function () {
    function MccScrollspyGroupDirective(mccScrollspyService) {
        this.mccScrollspyService = mccScrollspyService;
    }
    Object.defineProperty(MccScrollspyGroupDirective.prototype, "name", {
        /**
         * Name of the scrollspy group
         */
        set: function (name) {
            this._name = name;
        },
        enumerable: true,
        configurable: true
    });
    MccScrollspyGroupDirective.prototype.ngAfterContentInit = function () {
        // add the group items
        var items = this.items.map(function (item) { return item; });
        this.mccScrollspyService.create(this._name, items);
    };
    tslib_1.__decorate([
        ContentChildren(MccScrollspyItemDirective),
        tslib_1.__metadata("design:type", QueryList)
    ], MccScrollspyGroupDirective.prototype, "items", void 0);
    tslib_1.__decorate([
        Input('mccScrollspyGroup'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], MccScrollspyGroupDirective.prototype, "name", null);
    MccScrollspyGroupDirective = tslib_1.__decorate([
        Directive({
            selector: '[mccScrollspyGroup], [mcc-scrollspy-group]',
            exportAs: 'mccScrollspyGroup',
        }),
        tslib_1.__metadata("design:paramtypes", [MccScrollspyService])
    ], MccScrollspyGroupDirective);
    return MccScrollspyGroupDirective;
}());
export { MccScrollspyGroupDirective };
