// material-community-components/esm5/scrollspy/scrollspy.service.js
import * as tslib_1 from "tslib";
import { Injectable, Inject } from '@angular/core';
import { SCROLLSPY_ANIMATION_SMOOTH } from './scrollspy';
import { BehaviorSubject, fromEvent } from 'rxjs';
import { debounceTime, withLatestFrom } from 'rxjs/operators';
var MccScrollspyService = /** @class */ (function () {
    function MccScrollspyService(window) {
        var _this = this;
        this.window = window;
        /**
         * When scroll is from click event, change this attr to true
         * So scroll event obeservable doesn't emit any update
         */
        this._fromClick = false;
        /**
         * List of scrollspy group
         */
        this.data = [];
        // listen to scroll event
        this._scrollSub = fromEvent(window, 'scroll')
            .pipe(debounceTime(50), withLatestFrom(function () { return window.scrollY; }))
            .subscribe(function (position) {
            if (!_this._fromClick) {
                _this._updateFocused(position);
            }
            _this._fromClick = false;
        });
    }
    MccScrollspyService.prototype.ngOnDestroy = function () {
        if (this._scrollSub && !this._scrollSub.closed) {
            this._scrollSub.unsubscribe();
        }
    };
    /**
     * Update information about wich element is on focus
     * @param position number
     */
    MccScrollspyService.prototype._updateFocused = function (position) {
        this.data.forEach(function (group) {
            var items = [];
            group.items.getValue().forEach(function (item, index) {
                item.focus = false;
                if (item.top <= position) {
                    if (items[index - 1]) {
                        items[index - 1].focus = false;
                    }
                    item.focus = true;
                }
                items.push(item);
            });
            group.items.next(items);
        });
    };
    /**
     * Create new group of items
     * @param name string
     * @param items MccScrollspyItemDirective[]
     * @param animation ScrollBehavior
     */
    MccScrollspyService.prototype.create = function (name, items, animation) {
        var group = this.data.find(function (group) { return group.name === name; });
        if (!group) {
            group = {
                name: name,
                animation: animation || SCROLLSPY_ANIMATION_SMOOTH,
                items: new BehaviorSubject(items || []),
            };
            this.data.push(group);
        }
        else {
            group.items.next(items || []);
        }
        return group;
    };
    /**
     * Return observable of the group
     * @param name string
     */
    MccScrollspyService.prototype.group = function (name) {
        var group = this.data.find(function (g) { return g.name === name; });
        if (!group) {
            group = this.create(name);
        }
        return group.items.asObservable();
    };
    /**
     * Scroll to one of the items
     * @param name string
     * @param id string
     */
    MccScrollspyService.prototype.scrollTo = function (name, id) {
        var _this = this;
        var group = this.data.find(function (group) { return group.name === name; });
        group.items.getValue().forEach(function (item) {
            item.focus = false;
            if (item.id === id) {
                _this._fromClick = true;
                _this._updateFocused(item.top);
                window.scrollTo({ top: item.top, behavior: group.animation });
            }
        });
    };
    MccScrollspyService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__param(0, Inject('Window')),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], MccScrollspyService);
    return MccScrollspyService;
}());
export { MccScrollspyService };
