// material-community-components/esm2015/color-picker/color-picker.directives.js
import * as tslib_1 from "tslib";
var MccColorPickerOriginDirective_1;
import { ChangeDetectorRef, Directive, ElementRef, forwardRef, Input, Inject, Output, Renderer2, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { MccColorPickerComponent } from './color-picker.component';
import { EMPTY_COLOR, coerceHexaColor, isValidColor } from './color-picker';
import { BehaviorSubject } from 'rxjs';
/**
 * This directive change the background of the button
 */
let MccColorPickerOptionDirective = class MccColorPickerOptionDirective {
    constructor(elementRef, render, emptyColor) {
        this.elementRef = elementRef;
        this.render = render;
        this.emptyColor = emptyColor;
        this._color = emptyColor;
    }
    /**
     * Receive the color
     */
    get color() {
        return this._color;
    }
    set color(value) {
        this._color = value;
    }
    ngAfterViewInit() {
        if (this.color) {
            let color;
            if (typeof this.color === 'string') {
                color = this.color;
            }
            else {
                color = this.color.value;
                this.render.setAttribute(this.elementRef.nativeElement, 'aria-label', this.color.text);
            }
            if (isValidColor(color)) {
                // apply the color
                this.render.setStyle(this.elementRef.nativeElement, 'background', coerceHexaColor(color) || this.emptyColor);
            }
        }
    }
};
tslib_1.__decorate([
    Input('mccColorPickerOption'),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], MccColorPickerOptionDirective.prototype, "color", null);
MccColorPickerOptionDirective = tslib_1.__decorate([
    Directive({
        selector: '[mccColorPickerOption], [mcc-color-picker-option]',
        exportAs: 'mccColorPickerOption',
    }),
    tslib_1.__param(2, Inject(EMPTY_COLOR)),
    tslib_1.__metadata("design:paramtypes", [ElementRef,
        Renderer2, String])
], MccColorPickerOptionDirective);
export { MccColorPickerOptionDirective };
/**
 * Directive applied to an element to make it usable as an origin for an ColorPicker.
 */
let MccColorPickerOriginDirective = MccColorPickerOriginDirective_1 = class MccColorPickerOriginDirective {
    /**
     * Reference to the element on which the directive is applied.
     */
    constructor(elementRef, renderer, emptyColor) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.emptyColor = emptyColor;
        /**
         * Emit changes from the origin
         */
        this.change = new BehaviorSubject('');
        // listen changes onkeyup and update color picker
        renderer.listen(elementRef.nativeElement, 'keyup', (event) => {
            const value = event.currentTarget['value'];
            if (event.isTrusted && isValidColor(value)) {
                this.writeValueFromKeyup(coerceHexaColor(value) || this.emptyColor);
            }
        });
    }
    /**
     * This method will be called by the forms API to write to the view when
     * programmatic (model -> view) changes are requested.
     */
    writeValue(color) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'value', color);
        this.change.next(color);
        if (this.propagateChanges) {
            this.propagateChanges(color);
        }
    }
    /**
     * This method will be called by the color picker
     */
    writeValueFromColorPicker(color) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'value', color);
        this.propagateChanges(color);
    }
    /**
     * This method will be called from origin whe key is up
     */
    writeValueFromKeyup(color) {
        this.change.next(color);
        this.propagateChanges(color);
    }
    /**
     * This is called by the forms API on initialization so it can update the
     * form model when values propagate from the view (view -> model).
     * @param fn any
     */
    registerOnChange(fn) {
        this.propagateChanges = fn;
    }
    /**
     * This is called by the forms API on initialization so it can update the form model on blur
     * @param fn any
     */
    registerOnTouched(fn) { }
    /**
     * called by the forms API when the control status changes to or from "DISABLED"
     * @param isDisabled boolean
     */
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    }
};
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", BehaviorSubject)
], MccColorPickerOriginDirective.prototype, "change", void 0);
MccColorPickerOriginDirective = MccColorPickerOriginDirective_1 = tslib_1.__decorate([
    Directive({
        selector: '[mcc-color-picker-origin], [mccColorPickerOrigin]',
        exportAs: 'mccColorPickerOrigin',
        providers: [
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => MccColorPickerOriginDirective_1),
                multi: true,
            },
        ],
    }),
    tslib_1.__param(2, Inject(EMPTY_COLOR)),
    tslib_1.__metadata("design:paramtypes", [ElementRef,
        Renderer2, String])
], MccColorPickerOriginDirective);
export { MccColorPickerOriginDirective };
/**
 * Directive connect an color picker with any input, select or textarea.
 * The color picker will be automatically updated when the value of the origin is
 * changed.
 */
let MccConnectedColorPickerDirective = class MccConnectedColorPickerDirective {
    constructor(colorPicker, changeDetectorRef, emptyColor) {
        this.colorPicker = colorPicker;
        this.changeDetectorRef = changeDetectorRef;
        this.emptyColor = emptyColor;
    }
    ngAfterViewInit() {
        if (!this._colorPickerSub) {
            this._attachColorPicker();
        }
    }
    ngOnDestroy() {
        if (this._colorPickerSub && !this._colorPickerSub.closed) {
            this._colorPickerSub.unsubscribe();
        }
        if (this._originSub && !this._originSub.closed) {
            this._originSub.unsubscribe();
        }
    }
    /**
     * Attach color picker and origin
     */
    _attachColorPicker() {
        // subscribe to origin change to update color picker
        this._originSub = this.origin.change.subscribe(value => {
            if (isValidColor(value) ||
                (value === this.emptyColor && this.colorPicker.selectedColor !== this.emptyColor)) {
                this.colorPicker.updateTmpSelectedColor(value);
            }
            this.colorPicker.selectedColor = value;
            this.changeDetectorRef.detectChanges();
        });
        // subscribe to color picker changes and set on origin element
        this._colorPickerSub = this.colorPicker.change.subscribe(value => this.origin.writeValueFromColorPicker(value));
    }
};
tslib_1.__decorate([
    Input('mccConnectedColorPickerOrigin'),
    tslib_1.__metadata("design:type", MccColorPickerOriginDirective)
], MccConnectedColorPickerDirective.prototype, "origin", void 0);
MccConnectedColorPickerDirective = tslib_1.__decorate([
    Directive({
        selector: '[mcc-connected-color-picker], [mccConnectedColorPicker]',
        exportAs: 'mccConnectedColorPicker',
    }),
    tslib_1.__param(2, Inject(EMPTY_COLOR)),
    tslib_1.__metadata("design:paramtypes", [MccColorPickerComponent,
        ChangeDetectorRef, String])
], MccConnectedColorPickerDirective);
export { MccConnectedColorPickerDirective };
