// material-community-components/esm5/color-picker/color-picker.service.js
import * as tslib_1 from "tslib";
import { Injectable, Inject } from '@angular/core';
import { coerceHexaColor, isValidColor, EMPTY_COLOR, USED_COLORS } from './color-picker';
import { BehaviorSubject } from 'rxjs';
var MccColorPickerService = /** @class */ (function () {
    function MccColorPickerService(emptyColor, usedColors) {
        this.emptyColor = emptyColor;
        this.usedColors = usedColors;
        /**
         * Array of all used colors
         */
        this._colors = new BehaviorSubject([]);
        this._colors.next(usedColors);
    }
    /**
     * Add new color to used colors
     * @param color string
     */
    MccColorPickerService.prototype.addColor = function (color) {
        if (!color || !isValidColor(color)) {
            return;
        }
        color = coerceHexaColor(color) || this.emptyColor;
        var colors = this._colors.getValue();
        if (!colors.find(function (_color) { return _color === color; })) {
            colors.push(color);
            this._colors.next(colors);
        }
    };
    /**
     * Return Observable of colors
     */
    MccColorPickerService.prototype.getColors = function () {
        return this._colors.asObservable();
    };
    /**
     * Reset the array of used colors
     */
    MccColorPickerService.prototype.resetUseColors = function () {
        this._colors.next([]);
    };
    MccColorPickerService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__param(0, Inject(EMPTY_COLOR)),
        tslib_1.__param(1, Inject(USED_COLORS)),
        tslib_1.__metadata("design:paramtypes", [String, Array])
    ], MccColorPickerService);
    return MccColorPickerService;
}());
export { MccColorPickerService };
