// material-community-components/esm5/color-picker/color-picker.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { MatButtonModule, MatFormFieldModule, MatInputModule } from '@angular/material';
import { EMPTY_COLOR, USED_COLORS } from './color-picker';
import { MccColorPickerService } from './color-picker.service';
import { MccColorPickerComponent } from './color-picker.component';
import { MccColorPickerSelectorComponent } from './color-picker-selector.component';
import { MccColorPickerCollectionComponent } from './color-picker-collection.component';
import { MccConnectedColorPickerDirective, MccColorPickerOriginDirective, MccColorPickerOptionDirective, } from './color-picker.directives';
var ɵ0 = [];
var MccColorPickerModule = /** @class */ (function () {
    function MccColorPickerModule() {
    }
    MccColorPickerModule_1 = MccColorPickerModule;
    /**
     *
     */
    MccColorPickerModule.forRoot = function (config) {
        return {
            ngModule: MccColorPickerModule_1,
            providers: [
                { provide: EMPTY_COLOR, useValue: config.empty_color || 'none' },
                { provide: USED_COLORS, useValue: config.used_colors || [] }
            ],
        };
    };
    var MccColorPickerModule_1;
    MccColorPickerModule = MccColorPickerModule_1 = tslib_1.__decorate([
        NgModule({
            imports: [
                CommonModule,
                PortalModule,
                OverlayModule,
                ReactiveFormsModule,
                MatButtonModule,
                MatFormFieldModule,
                MatInputModule,
            ],
            declarations: [
                MccColorPickerComponent,
                MccConnectedColorPickerDirective,
                MccColorPickerSelectorComponent,
                MccColorPickerOriginDirective,
                MccColorPickerOptionDirective,
                MccColorPickerCollectionComponent,
            ],
            exports: [
                MccColorPickerComponent,
                MccConnectedColorPickerDirective,
                MccColorPickerOriginDirective,
                MccColorPickerCollectionComponent,
            ],
            providers: [
                MccColorPickerService,
                { provide: EMPTY_COLOR, useValue: 'none' },
                { provide: USED_COLORS, useValue: ɵ0 }
            ],
        })
    ], MccColorPickerModule);
    return MccColorPickerModule;
}());
export { MccColorPickerModule };
export { ɵ0 };
