// material-community-components/esm2015/timer-picker/timer-picker.directives.js
import * as tslib_1 from "tslib";
var MccTimerPickerOriginDirective_1;
import { ChangeDetectorRef, Directive, ElementRef, forwardRef, Input, Output, Renderer2, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { MccTimerPickerComponent } from './timer-picker.component';
import { BehaviorSubject } from 'rxjs';
let MccTimerPickerOriginDirective = MccTimerPickerOriginDirective_1 = class MccTimerPickerOriginDirective {
    /**
     * Reference to the element on which the directive is applied.
     */
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        /**
         * Emit changes from the origin
         */
        this.change = new BehaviorSubject('');
        /**
         * Emit changes from the origin
         */
        this.hasFocus = new BehaviorSubject(false);
        // listen focus
        renderer.listen(elementRef.nativeElement, 'focus', () => this.hasFocus.next(true));
    }
    /**
     * This method will be called by the forms API to write to the view when
     * programmatic (model -> view) changes are requested.
     */
    writeValue(time) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'value', time);
        this.change.next(time);
        if (this.propagateChanges) {
            this.propagateChanges(time);
        }
    }
    /**
     * This method will be called by the time picker
     */
    writeValueFromTimerPicker(time) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'value', time);
        this.propagateChanges(time);
    }
    /**
     * This method will be called from origin whe key is up
     */
    writeValueFromKeyup(time) {
        this.change.next(time);
        this.propagateChanges(time);
    }
    /**
     * This is called by the forms API on initialization so it can update the
     * form model when values propagate from the view (view -> model).
     * @param fn any
     */
    registerOnChange(fn) {
        this.propagateChanges = fn;
    }
    /**
     * This is called by the forms API on initialization so it can update the form model on blur
     * @param fn any
     */
    registerOnTouched(fn) { }
    /**
     * called by the forms API when the control status changes to or from "DISABLED"
     * @param isDisabled boolean
     */
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    }
};
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", BehaviorSubject)
], MccTimerPickerOriginDirective.prototype, "change", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", BehaviorSubject)
], MccTimerPickerOriginDirective.prototype, "hasFocus", void 0);
MccTimerPickerOriginDirective = MccTimerPickerOriginDirective_1 = tslib_1.__decorate([
    Directive({
        selector: '[mccTimerPickerOrigin], [mcc-timer-picker-origin]',
        exportAs: 'mccTimerPickerOrigin',
        providers: [
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => MccTimerPickerOriginDirective_1),
                multi: true,
            },
        ],
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], MccTimerPickerOriginDirective);
export { MccTimerPickerOriginDirective };
let MccConnectedTimerPickerDirective = class MccConnectedTimerPickerDirective {
    constructor(timerPicker, changeDetectorRef) {
        this.timerPicker = timerPicker;
        this.changeDetectorRef = changeDetectorRef;
        this.timerPicker.connected = true;
    }
    ngAfterViewInit() {
        if (!this._timerPickerSub) {
            this.timerPicker.trigger = this.origin;
            this._attachTimerPicker();
        }
    }
    ngOnDestroy() {
        if (this._originFocus && !this._originFocus.closed) {
            this._originFocus.unsubscribe();
        }
        if (this._timerPickerSub && !this._timerPickerSub.closed) {
            this._timerPickerSub.unsubscribe();
        }
    }
    /**
     * Attach the timer picker to origin element (input)
     */
    _attachTimerPicker() {
        this._originFocus = this.origin.hasFocus.subscribe(focused => {
            this.timerPicker.focus = 'hour';
            this.timerPicker.isOpen = focused;
            this.changeDetectorRef.detectChanges();
        });
        this._timerPickerSub = this.timerPicker.selected.subscribe(value => this.origin.writeValueFromTimerPicker(value));
    }
};
tslib_1.__decorate([
    Input('mccConnectedTimerPickerOrigin'),
    tslib_1.__metadata("design:type", MccTimerPickerOriginDirective)
], MccConnectedTimerPickerDirective.prototype, "origin", void 0);
MccConnectedTimerPickerDirective = tslib_1.__decorate([
    Directive({
        selector: '[mccConnectedTimerPicker], [mcc-connected-timer-picker]',
        exportAs: 'mccConnectedTimerPicker',
    }),
    tslib_1.__metadata("design:paramtypes", [MccTimerPickerComponent,
        ChangeDetectorRef])
], MccConnectedTimerPickerDirective);
export { MccConnectedTimerPickerDirective };
