// material-community-components/esm5/speed-dial/speed-dial.component.js
import * as tslib_1 from "tslib";
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Component, ContentChild, EventEmitter, Input, Output } from '@angular/core';
import { SPIN_ANIMATION } from './animations';
import { MccSpeedDialActionsComponent } from './speed-dial-actions.component';
var MccSpeedDialComponent = /** @class */ (function () {
    function MccSpeedDialComponent() {
        this._isOpen = false;
        this._hover = false;
        this._spin = true;
        this._direction = 'up';
        /**
         * Event emitted when open state change
         */
        this.openStateChange = new EventEmitter();
    }
    Object.defineProperty(MccSpeedDialComponent.prototype, "isOpen", {
        get: function () {
            return this._isOpen;
        },
        /**
         * Set initial 'open' state
         */
        set: function (open) {
            this._isOpen = coerceBooleanProperty(open);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MccSpeedDialComponent.prototype, "hover", {
        /**
         * When enabled, handle open/close state on mouse hover
         */
        set: function (hover) {
            this._hover = coerceBooleanProperty(hover);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MccSpeedDialComponent.prototype, "spin", {
        get: function () {
            return this._spin;
        },
        /**
         * Enable/disable spin animation when button is clicked or hovered
         */
        set: function (spin) {
            this._spin = spin;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MccSpeedDialComponent.prototype, "direction", {
        get: function () {
            return this._direction;
        },
        /**
         * Define the direction of the actions button
         * Directions available are: up | down | left | right
         */
        set: function (direction) {
            this._direction = direction;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Call fab speed dial actions functions to change the
     * visibility of the buttons
     */
    MccSpeedDialComponent.prototype._setActionsState = function () {
        if (this._isOpen) {
            this.actions.show(this._direction);
        }
        else {
            this.actions.hide(this._direction);
        }
    };
    /**
     * Set initial state to the action buttons inside speed-dial-actions
     */
    MccSpeedDialComponent.prototype.ngAfterViewInit = function () {
        this._setActionsState();
    };
    /**
     *
     */
    MccSpeedDialComponent.prototype.ngOnChanges = function (changes) {
        if ('isOpen' in changes && changes['isOpen'].previousValue !== undefined) {
            this._setActionsState();
        }
    };
    /**
     * When mouseHover is enabled and state is closed
     * calls toggle to open the actions
     */
    MccSpeedDialComponent.prototype.hoverStart = function () {
        if (this._hover && !this._isOpen) {
            this.toggle();
        }
    };
    /**
     * When mouseHover is enabled and state is open
     * calls toggle to close the actions
     */
    MccSpeedDialComponent.prototype.hoverStop = function () {
        if (this._hover && this._isOpen) {
            this.toggle();
        }
    };
    /**
     * Change the open state
     */
    MccSpeedDialComponent.prototype.toggle = function () {
        this._isOpen = !this._isOpen;
        this._setActionsState();
        this.openStateChange.emit(this._isOpen);
    };
    tslib_1.__decorate([
        ContentChild(MccSpeedDialActionsComponent, { static: false }),
        tslib_1.__metadata("design:type", MccSpeedDialActionsComponent)
    ], MccSpeedDialComponent.prototype, "actions", void 0);
    tslib_1.__decorate([
        Input('open'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], MccSpeedDialComponent.prototype, "isOpen", null);
    tslib_1.__decorate([
        Input('mouseHover'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], MccSpeedDialComponent.prototype, "hover", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], MccSpeedDialComponent.prototype, "spin", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], MccSpeedDialComponent.prototype, "direction", null);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], MccSpeedDialComponent.prototype, "openStateChange", void 0);
    MccSpeedDialComponent = tslib_1.__decorate([
        Component({
            selector: 'mcc-speed-dial',
            template: "<div class=\"mcc-speed-dial-container\" [ngClass]=\"'mcc-speed-dial-direction-' + direction\">\n  <button mat-fab [@spin]=\"spin && isOpen ? 'open' : 'closed'\" class=\"mat-elevation-z1\" (mouseenter)=\"hoverStart()\"\n    (mouseleave)=\"hoverStop()\" (click)=\"toggle()\">\n    <ng-content></ng-content>\n  </button>\n\n  <ng-content select=\"mcc-speed-dial-actions\"></ng-content>\n</div>",
            animations: [SPIN_ANIMATION],
            styles: [".mcc-speed-dial-container{position:relative;display:flex;align-items:center;z-index:20}.mcc-speed-dial-container button{pointer-events:auto;z-index:24}.mcc-speed-dial-container ::ng-deep mcc-speed-dial-actions{display:flex;height:auto}.mcc-speed-dial-container.mcc-speed-dial-direction-up{flex-direction:column}.mcc-speed-dial-container.mcc-speed-dial-direction-up button{order:2}.mcc-speed-dial-container.mcc-speed-dial-direction-up ::ng-deep mcc-speed-dial-actions{flex-direction:column-reverse;order:1}.mcc-speed-dial-container.mcc-speed-dial-direction-up ::ng-deep mcc-speed-dial-actions .mat-mini-fab{margin-bottom:10px}.mcc-speed-dial-container.mcc-speed-dial-direction-down{flex-direction:column}.mcc-speed-dial-container.mcc-speed-dial-direction-down button{order:1}.mcc-speed-dial-container.mcc-speed-dial-direction-down ::ng-deep mcc-speed-dial-actions{flex-direction:column;order:2}.mcc-speed-dial-container.mcc-speed-dial-direction-down ::ng-deep mcc-speed-dial-actions .mat-mini-fab{margin-top:10px}.mcc-speed-dial-container.mcc-speed-dial-direction-left{flex-direction:row}.mcc-speed-dial-container.mcc-speed-dial-direction-left button{order:2}.mcc-speed-dial-container.mcc-speed-dial-direction-left ::ng-deep mcc-speed-dial-actions{flex-direction:row-reverse;order:1}.mcc-speed-dial-container.mcc-speed-dial-direction-left ::ng-deep mcc-speed-dial-actions .mat-mini-fab{margin-right:10px}.mcc-speed-dial-container.mcc-speed-dial-direction-right{flex-direction:row}.mcc-speed-dial-container.mcc-speed-dial-direction-right button{order:1}.mcc-speed-dial-container.mcc-speed-dial-direction-right ::ng-deep mcc-speed-dial-actions{flex-direction:row;order:2}.mcc-speed-dial-container.mcc-speed-dial-direction-right ::ng-deep mcc-speed-dial-actions .mat-mini-fab{margin-left:10px}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], MccSpeedDialComponent);
    return MccSpeedDialComponent;
}());
export { MccSpeedDialComponent };
