// material-community-components/esm5/speed-dial/speed-dial-actions.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChildren, Input, QueryList, Renderer2, ViewEncapsulation } from '@angular/core';
import { MatButton } from '@angular/material';
import { BehaviorSubject } from 'rxjs';
import { Z_INDEX } from './animations';
var MccSpeedDialActionsComponent = /** @class */ (function () {
    function MccSpeedDialActionsComponent(renderer) {
        this.renderer = renderer;
        this._animation = new BehaviorSubject('scale');
    }
    Object.defineProperty(MccSpeedDialActionsComponent.prototype, "animation", {
        /**
         * Set type of animation will be executed on open/close
         * Type available are: scale | fling
         */
        set: function (animation) {
            this._animation.next(animation);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * The z-index style and animation class are handle separate because
     * z-index will be set only one time, and the animation class will be set
     * every time the animation change
     */
    MccSpeedDialActionsComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        // set z-index style to each button action
        this._buttons.forEach(function (button, index) {
            _this.renderer.setStyle(button._elementRef.nativeElement, 'z-index', (Z_INDEX - index));
        });
        // set the animation class to each button action
        this._animation.subscribe(function (animation) {
            var nextAnimationClass = "speed-dial-item-animation-" + animation;
            _this._buttons.forEach(function (button) {
                if (_this._lastAnimationClass) {
                    _this.renderer.removeClass(button._elementRef.nativeElement, _this._lastAnimationClass);
                }
                _this.renderer.addClass(button._elementRef.nativeElement, nextAnimationClass);
            });
            _this._lastAnimationClass = nextAnimationClass;
        });
    };
    /**
     * Responsible for change the state of the action buttons to visible
     *
     * @param direction DIRECTION
     */
    MccSpeedDialActionsComponent.prototype.show = function (direction) {
        var _this = this;
        switch (this._animation.value) {
            case 'scale': {
                this._buttons.forEach(function (button, index) {
                    var transition = 3 + (65 * index);
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'transition-delay', transition + "ms");
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'opacity', '1');
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'transform', 'scale(1)');
                });
                break;
            }
            case 'fling': {
                var translateFn_1 = (direction == 'up' || direction == 'down') ? 'translateY' : 'translateX';
                var sign_1 = (direction == 'down' || direction == 'right') ? '-' : '';
                this._buttons.forEach(function (button) {
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'transition-delay', '0ms');
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'opacity', '1');
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'transform', translateFn_1 + "(" + sign_1 + "0)");
                });
            }
        }
    };
    /**
     * Hide all the buttons action
     *
     * @param direction DIRECTION
     */
    MccSpeedDialActionsComponent.prototype.hide = function (direction) {
        var _this = this;
        switch (this._animation.value) {
            case 'scale': {
                this._buttons.forEach(function (button, index) {
                    var transition = 3 - (65 * index);
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'transition-delay', transition + "ms");
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'opacity', '0');
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'transform', 'scale(0)');
                });
                break;
            }
            case 'fling': {
                var translateFn_2 = (direction == 'up' || direction == 'down') ? 'translateY' : 'translateX';
                var sign_2 = (direction == 'down' || direction == 'right') ? '-' : '';
                this._buttons.forEach(function (button, index) {
                    var transform = (55 * (index + 1) - (index * 5));
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'transition-delay', '0ms');
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'opacity', '1');
                    _this.renderer.setStyle(button._elementRef.nativeElement, 'transform', translateFn_2 + "(" + sign_2 + transform + "px)");
                });
            }
        }
    };
    tslib_1.__decorate([
        ContentChildren(MatButton),
        tslib_1.__metadata("design:type", QueryList)
    ], MccSpeedDialActionsComponent.prototype, "_buttons", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], MccSpeedDialActionsComponent.prototype, "animation", null);
    MccSpeedDialActionsComponent = tslib_1.__decorate([
        Component({
            selector: 'mcc-speed-dial-actions',
            template: "<ng-content select=\"button\"></ng-content>",
            encapsulation: ViewEncapsulation.None,
            styles: ["mcc-speed-dial-actions .speed-dial-item-animation-scale{transform:scale(0);transition:.3s cubic-bezier(.55,0,.55,.2);transition-duration:.14286s}mcc-speed-dial-actions .speed-dial-item-animation-fling{display:block;opacity:1;transition:.3s cubic-bezier(.55,0,.55,.2)}"]
        }),
        tslib_1.__metadata("design:paramtypes", [Renderer2])
    ], MccSpeedDialActionsComponent);
    return MccSpeedDialActionsComponent;
}());
export { MccSpeedDialActionsComponent };
